subcommand/server-acl-init/connect_inject.go

```diff
@@ -152,16 +152,20 @@ func (c *Command) createAuthMethodTmpl(authMethodName string) (api.ACLAuthMethod
 
 	// ServiceAccounts always have a secret name. The secret
 	// contains the JWT token.
-	saSecretName := authMethodServiceAccount.Secrets[0].Name
-
-	// Get the secret that will contain the ServiceAccount JWT token.
+	// Because there could be multiple secrets attached to the service account,
+	// we need pick the first one of type "kubernetes.io/service-account-token".
 	var saSecret *apiv1.Secret
-	err = c.untilSucceeds(fmt.Sprintf("getting %s Secret", saSecretName),
-		func() error {
-			var err error
-			saSecret, err = c.clientset.CoreV1().Secrets(c.flagK8sNamespace).Get(context.TODO(), saSecretName, metav1.GetOptions{})
-			return err
-		})
+	for _, secret := range authMethodServiceAccount.Secrets {
+		err = c.untilSucceeds(fmt.Sprintf("getting %s Secret", secret.Name),
+			func() error {
+				var err error
+				saSecret, err = c.clientset.CoreV1().Secrets(c.flagK8sNamespace).Get(context.TODO(), secret.Name, metav1.GetOptions{})
+				return err
+			})
+		if saSecret.Type == apiv1.SecretTypeServiceAccountToken {
+			break
+		}
+	}
```

Comment:
I know this issue existed with the previous implementation but while we're in here what do you think about doing something like this. As written if we don't find a secret of type ServiceAccountToken then we're going to use the last secret in the list.

```suggestion
	for _, secretRef := range authMethodServiceAccount.Secrets {
	        var secret *apiv1.Secret
		err = c.untilSucceeds(fmt.Sprintf("getting %s Secret", secretRef.Name),
			func() error {
				var err error
				secret, err = c.clientset.CoreV1().Secrets(c.flagK8sNamespace).Get(context.TODO(), secretRef.Name, metav1.GetOptions{})
				return err
			})
		if secret.Type == apiv1.SecretTypeServiceAccountToken {
			saSecret = secret
			break
		}
	}
	if err != nil {
	   return api.ACLAuthMethod{}, err
	}
	if saSecret == nil {
	  return api.ACLAuthMethod{}, fmt.Errorf("Found no secret of type ....")
	}
```

Reply:
Will that be a valid secret? Wouldn't it be preferable to error out and say no valid secret found?

Reply:
>Will that be a valid secret? Wouldn't it be preferable to error out and say no valid secret found?

Is this a question for me or did we just have overlapping reviews? :D

Reply:
lol. Definitely an overlapping review :P Sorry for the confusion.

Reply:
Yeah definitely, good catch!